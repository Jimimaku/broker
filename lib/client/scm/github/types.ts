export type GitHubCommitPayload = {
  message: string;
  tree: string;
  parents: string[];
  author: GitUser;
  committer: GitUser;
  signature?: string;
};

export type GitUser = {
  name: string;
  email: string;
  date: Date;
};

export type GitHubCreateTreePayload = {
  owner: string;
  repo: string;
  base_tree: string;
  tree: GitHubTree[];
};

export type GitHubTree = {
  content: string;
  mode: '040000' | '100644' | '100755' | '120000' | '160000';
  path: string;
  type: 'blob' | 'commit' | 'tree';
};
