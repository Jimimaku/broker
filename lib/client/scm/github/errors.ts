export class GitHubCommitParsingError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, GitHubCommitParsingError.prototype);
  }
}
