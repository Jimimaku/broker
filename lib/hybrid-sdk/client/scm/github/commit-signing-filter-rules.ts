import type { FilterRule } from '../types';

export function getCommitSigningGitHubFilterRules(): FilterRule[] {
  const origin = 'https://${GITHUB_TOKEN}@${GITHUB_API}';

  return [
    {
      '//': 'create tree',
      method: 'POST',
      path: '/repos/:name/:repo/git/trees',
      origin,
    },
    {
      '//': 'create tree refs',
      method: 'POST',
      path: '/repos/:name/:repo/git/trees/:ref',
      origin,
    },
    {
      '//': 'create commit',
      method: 'POST',
      path: '/repos/:name/:repo/git/commits',
      origin,
    },
    {
      '//': 'update pull request',
      method: 'PATCH',
      path: '/repos/:name/:repo/pulls/:pull_number',
      origin,
    },
    {
      '//': 'delete ref head',
      method: 'DELETE',
      path: '/repos/:name/:repo/git/refs/heads/:branch_name',
      origin,
    },
  ];
}
