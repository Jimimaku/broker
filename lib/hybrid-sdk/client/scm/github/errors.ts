export class GitHubCommitParsingError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, GitHubCommitParsingError.prototype);
  }
}

export class GitHubTreeParsingError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, GitHubTreeParsingError.prototype);
  }
}

export class GitHubTreeValidationError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, GitHubTreeValidationError.prototype);
  }
}
