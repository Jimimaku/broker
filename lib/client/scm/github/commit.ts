import * as logger from '../../../log';
import { GitHubCommitParsingError } from './errors';
import type { GitHubCommitPayload, GitUser } from './types';

export function isGitHubCreateCommitEndpoint(options: {
  method: string;
  url: string;
}): boolean {
  // https://docs.github.com/en/rest/git/commits?apiVersion=2022-11-28#create-a-commit
  return options.method === 'POST' && options.url.includes('/git/commits');
}

export function convertBodyToGitHubCommitPayload(
  body: unknown,
  options: {
    committerName: string;
    committerEmail: string;
  },
): GitHubCommitPayload {
  const bodyAsString = convertBodyToStringIfNeeded(body);

  try {
    const bodyAsJson = JSON.parse(bodyAsString);
    return {
      message: bodyAsJson.message,
      tree: bodyAsJson.tree,
      parents: bodyAsJson.parents as [],
      author: {
        name: bodyAsJson.author.name,
        email: bodyAsJson.author.email,
        date: convertToDate(bodyAsJson.author.date),
      },
      committer: {
        name: options.committerName,
        email: options.committerEmail,
        date: convertToDate(bodyAsJson.committer?.date || new Date()),
      },
    } satisfies GitHubCommitPayload;
  } catch (error) {
    logger.error({ error }, 'Could not parse GitHub commit payload body');
    throw new GitHubCommitParsingError(
      'Could not parse GitHub commit payload body',
    );
  }
}

const convertBodyToStringIfNeeded = (body: unknown): string => {
  if (isUint8Array(body)) {
    return Buffer.from(body).toString();
  } else if (typeof body === 'string' || body instanceof String) {
    return body as string;
  } else {
    throw new Error('body must be string or Uint8Array');
  }
};

const isUint8Array = (data: unknown): data is Uint8Array => {
  return !!(data && data instanceof Uint8Array);
};

const convertToDate = (input: unknown): Date => {
  if (input && typeof input === 'string') {
    return new Date(input);
  } else if (input && input instanceof Date) {
    return input;
  } else {
    throw new Error('could not convert input to date');
  }
};

export function stringifyGitHubCommitPayload(
  commit: GitHubCommitPayload,
): string {
  return `tree ${commit.tree}
parent ${commit.parents[0]}
author ${stringifyGitUser(commit.author)}
committer ${stringifyGitUser(commit.committer)}

${commit.message}`.trim();
}

const stringifyGitUser = (user: GitUser): string => {
  const timestamp = Math.floor(user.date.getTime() / 1000);
  return `${user.name} <${user.email}> ${timestamp} +0000`;
};
