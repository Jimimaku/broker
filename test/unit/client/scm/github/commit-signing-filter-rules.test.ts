import { getCommitSigningGitHubFilterRules } from '../../../../../lib/hybrid-sdk/client/scm/github/commit-signing-filter-rules';

describe('client/scm/github/commit-signing-filter-rules.ts', () => {
  describe('getCommitSigningGitHubFilterRules()', () => {
    it('should load commit signing rules correctly', async () => {
      const rules = getCommitSigningGitHubFilterRules();

      expect(rules).toMatchSnapshot();
    });
  });
});
