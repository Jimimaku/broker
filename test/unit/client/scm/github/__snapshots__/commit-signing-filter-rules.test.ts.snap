// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client/scm/github/commit-signing-filter-rules.ts getCommitSigningGitHubFilterRules() should load commit signing rules correctly 1`] = `
Array [
  Object {
    "//": "create tree",
    "method": "POST",
    "origin": "https://\${GITHUB_TOKEN}@\${GITHUB_API}",
    "path": "/repos/:name/:repo/git/trees",
  },
  Object {
    "//": "create tree refs",
    "method": "POST",
    "origin": "https://\${GITHUB_TOKEN}@\${GITHUB_API}",
    "path": "/repos/:name/:repo/git/trees/:ref",
  },
  Object {
    "//": "create commit",
    "method": "POST",
    "origin": "https://\${GITHUB_TOKEN}@\${GITHUB_API}",
    "path": "/repos/:name/:repo/git/commits",
  },
  Object {
    "//": "update pull request",
    "method": "PATCH",
    "origin": "https://\${GITHUB_TOKEN}@\${GITHUB_API}",
    "path": "/repos/:name/:repo/pulls/:pull_number",
  },
  Object {
    "//": "delete ref head",
    "method": "DELETE",
    "origin": "https://\${GITHUB_TOKEN}@\${GITHUB_API}",
    "path": "/repos/:name/:repo/git/refs/heads/:branch_name",
  },
]
`;
